core/src/test/java/io/parsingdata/metal/TrampolineTest.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2013-2016 Netherlands Forensic Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.parsingdata.metal;
+
+import static io.parsingdata.metal.SafeTrampoline.complete;
+import static io.parsingdata.metal.SafeTrampoline.intermediate;
+
+import java.math.BigInteger;
+
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import io.parsingdata.metal.Trampoline.CompletedTrampoline;
+import io.parsingdata.metal.Trampoline.IntermediateTrampoline;
+
+public class TrampolineTest {
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    @Test
+    public void resultOnIntermediateTrampoline() {
+        thrown.expect(UnsupportedOperationException.class);
+        thrown.expectMessage("An IntermediateTrampoline does not have a result.");
+        ((IntermediateTrampoline<Integer>) () -> null).result();
+    }
+
+    @Test
+    public void nextOnFinalTrampoline() {
+        thrown.expect(UnsupportedOperationException.class);
+        thrown.expectMessage("A CompletedTrampoline does not have a next computation.");
+        ((CompletedTrampoline<Integer>) () -> 42).next();
+    }
+
+    @Test
+    public void resultOnIntermediateSafeTrampoline() {
+        thrown.expect(UnsupportedOperationException.class);
+        thrown.expectMessage("An IntermediateTrampoline does not have a result.");
+        ((SafeTrampoline.IntermediateTrampoline<Integer>) () -> null).result();
+    }
+
+    @Test
+    public void nextOnFinalSafeTrampoline() {
+        thrown.expect(UnsupportedOperationException.class);
+        thrown.expectMessage("A CompletedTrampoline does not have a next computation.");
+        ((SafeTrampoline.CompletedTrampoline<Integer>) () -> 42).next();
+    }
+
+    @Test
+    public void noStackOverflow() {
+        // The 100000th Fibonacci number has 20899 digits
+        Assert.assertEquals(20899, fibonacci(BigInteger.ZERO, BigInteger.ONE, 100000).computeResult().toString().length());
```

Comment:
I wonder what the average stack size has to be in order for a standard recursive method to complete without overflow :smile: